src/indexer.rs

```diff
@@ -0,0 +1,624 @@
+/// bookkeeping.
+///
+/// Note that this trait has `async` methods, which is why the type signatures
+/// are daunting. The intended way of implementing it is to use the
+/// `async_trait` macro like so
+///
+/// ```
+/// # use concordium_rust_sdk::{indexer::*, v2::{self, FinalizedBlockInfo, QueryResult}};
+/// use concordium_rust_sdk::indexer::async_trait;
+/// # struct MyIndexer;
+/// #[async_trait]
+/// impl Indexer for MyIndexer {
+///     type Context = ();
+///     type Data = ();
+///
+///     async fn on_connect<'a>(
+///         &mut self,
+///         endpoint: v2::Endpoint,
+///         client: &'a mut v2::Client,
+///     ) -> QueryResult<Self::Context> {
+///         unimplemented!("Implement me.")
+///     }
+///
+///     async fn on_finalized<'a>(
+///         &self,
+///         client: v2::Client,
+///         ctx: &'a Self::Context,
+///         fbi: FinalizedBlockInfo,
+///     ) -> QueryResult<Self::Data> {
+///         unimplemented!("Implement me.")
+///     }
+///
+///     async fn on_failure(
+///         &mut self,
+///         ep: v2::Endpoint,
+///         successive_failures: u64,
+///         err: String,
+///     ) -> bool {
+///         unimplemented!("Implement me.")
+///     }
+/// }
+/// ```
+
+pub trait Indexer {
+    /// The data that is retrieved upon connecting to the endpoint and supplied
+    /// to each call of [`on_finalized`](Self::on_finalized).
+    type Context: Send + Sync;
+    /// The data returned by the [`on_finalized`](Self::on_finalized) call.
+    type Data: Send + Sync;
+
+    /// Called when a new connection is established to the given endpoint.
+    /// The return value from this method is passed to each call of
+    /// [`on_finalized`](Self::on_finalized).
+    async fn on_connect<'a>(
+        &mut self,
+        endpoint: v2::Endpoint,
+        client: &'a mut v2::Client,
+    ) -> QueryResult<Self::Context>;
+
+    /// The main method of this trait. It is called for each finalized block
+    /// that the indexer discovers. Note that the indexer might call this
+    /// concurrently for multiple blocks at the same time to speed up indexing.
+    ///
+    /// This method is meant to return errors that are unexpected, and if it
+    /// does return an error the indexer will attempt to reconnect to the
+    /// next endpoint.
+    async fn on_finalized<'a>(
+        &self,
+        client: v2::Client,
+        ctx: &'a Self::Context,
+        fbi: FinalizedBlockInfo,
+    ) -> QueryResult<Self::Data>;
+
+    /// Called when either connecting to the node or querying the node fails.
+    /// The number of successive failures without progress is passed to the
+    /// method which should return whether to stop indexing ([`true`]) or not
+    /// ([`false`]).
+    ///
+    /// The `err` passed to this method will contain a description of the error.
+    async fn on_failure(
+        &mut self,
+        endpoint: v2::Endpoint,
+        successive_failures: u64,
+        err: String,
```

Comment:
**Suggestion (non-blocking):** Introduce an error type instead of error strings.

Reply:
Added